include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -1673,117 +1549,91 @@ struct __parallel_sort_submitter<__internal::__optional_kernel_name<_LeafSortNam
+    template <typename _ExecutionPolicy, typename _Range, typename _Compare>
     auto
-    operator()(_ExecutionPolicy&& __exec, _Range&& __rng, _Merge __merge, _Compare __comp) const
+    operator()(_ExecutionPolicy&& __exec, _Range&& __rng, _Compare __comp) const
     {
         using _Policy = typename ::std::decay<_ExecutionPolicy>::type;
         using _Tp = oneapi::dpl::__internal::__value_t<_Range>;
         using _Size = oneapi::dpl::__internal::__difference_t<_Range>;
 
-        _Size __n = __rng.size();
+        const ::std::size_t __n = __rng.size();
         assert(__n > 1);
 
-        _PRINT_INFO_IN_DEBUG_MODE(__exec);
-
-        // __leaf: size of a block to sort using algorithm suitable for small sequences
-        // __optimal_chunk: best size of a block to merge duiring a step of the merge sort algorithm
-        // The coefficients were found experimentally
-        _Size __leaf = 4;
-        _Size __optimal_chunk = 4;
-        if (__exec.queue().get_device().is_cpu())
-        {
-            __leaf = 16;
-            __optimal_chunk = 32;
-        }
-        // Assume powers of 2
-        assert((__leaf & (__leaf - 1)) == 0);
-        assert((__optimal_chunk & (__optimal_chunk - 1)) == 0);
-
-        const _Size __leaf_steps = ((__n - 1) / __leaf) + 1;
+        const ::std::uint16_t __chunk = 4;
+        const _Size __steps = oneapi::dpl::__internal::__dpl_ceiling_div(__n, __chunk);
 
         // 1. Perform sorting of the leaves of the merge sort tree
-        sycl::event __event1 = __exec.queue().submit([&](sycl::handler& __cgh) {
+        sycl::event __event1 = __exec.queue().submit([&, __comp](sycl::handler& __cgh) {
             oneapi::dpl::__ranges::__require_access(__cgh, __rng);
-            __cgh.parallel_for<_LeafSortName...>(sycl::range</*dim=*/1>(__leaf_steps),
-                                                 [=](sycl::item</*dim=*/1> __item_id) {
-                                                     const _Size __idx = __item_id.get_linear_id() * __leaf;
-                                                     const _Size __start = __idx;
-                                                     const _Size __end = sycl::min(__start + __leaf, __n);
-                                                     __leaf_sort_kernel()(__rng, __start, __end, __comp);
-                                                 });
-        });
+            __cgh.parallel_for<_LeafSortName...>(sycl::range</*dim=*/1>(__steps), [=](sycl::item</*dim=*/1> __item_id)
+            {
+                const ::std::uint32_t i_elem = __item_id.get_linear_id() * __chunk;
 
-        _Size __sorted = __leaf;
-        // Chunk size cannot be bigger than size of a sorted sequence
-        _Size __chunk = ::std::min(__leaf, __optimal_chunk);
+                // forwarding references allow binding of internal tuple of references with rvalue
+                for(::std::uint8_t  __i = 0; __i < 2; ++__i)
```

Comment:
Yes, I kept in mind leaf sorting for custom leaf size... But, for merge phase experiment shows that chunk is 4 is optimal (at least for PVC). So, we have to have leaf is 2, not more.

Reply:
If we had a variative leaf size, 2 and 4 at least, we would be able to avoid calling the copyback kernel. 

For example, if we sort 16 items with leaf=2, then we call log2(16/2) = 3 iterations of the main sort kernel and then the copyback kernel, because the data is in the temporary buffer. If the leaf was 4, then there would be log2(16/4)=2 iterations without calling the copyback kernel.

Reply:
"leaf" - is size of each sorted subrange.
"chunk" - grainsize for each merge in "merge_sort" loop. "chunk" is n1+n2  or leaf*2, because for merge sort n1==n2=leaf.

Reply:
So, I'm a bit confused here still.  Really there are 3 values:

- `leaf` - size of each initally sorted subrange by the first phase.
- `chunk` - grainsize for each merge in merge phase loop. `chunk` is `n1+n2` or `leaf*2`, because for merge sort `n1==n2=leaf`.   **(doubles with each iteration)**
- `iters_per_work_item` - in the merge section this is the number of elements merged by each work item.  Our merge phase is coordinating to resolve ALL merges for the iteration step of all ranges with a single workgroup. I think the only restriction here is that `iters_per_work_item <= chunk` as we don't want to split a work item to two separate merges.  

I would expect your optimal value of 4 for the merge phase is for `iters_per_work_item`, not `chunk` as defined above.  We don't get to control the size of `chunk`, as we have a number of iterations and it `chunk` doubles at each one.

Unless something above is wrong, I think we can freely pick `leaf` size for "powers of 2"  `>= 2`, right?

We can still have a `iters_per_work_item = 4` for all iterations of the merge phase loop with `chunk` size going from `2 * leaf` -> `bit_ceil(__n)`. If we double the size of the `leaf`, we skip an iteration of the merge phase loop.  Obviously, you need to make sure the data ends up in the correct buffer, but otherwise I think it should be a benefit.  Selecting a leaf size to avoid an extra copy here makes sense to me.

Reply:
I update a little bit the definition of the variables.
`const ::std::uint32_t __leaf = __is_cpu ? 16 : 2;`
and 
`const ::std::uint32_t __chunk = __is_cpu ? 32 : 4;`

_chunk - is a constant for all merge iterations. It number of serial merging steps, which each work item does.

Reply:
OK, I think that makes sense.  
I believe that the leaf size can be any power of 2 greater than 2.  Do you disagree?

I have not checked performance, but I would guess that we could get benefit from larger leaf.

I also think that @dmitriy-sobolev 's suggestion is a good one for performance, but that would be another doubling of the number of kernels required.    Perhaps it makes more sense to add this to a future work issue , and plan to try to do it with specialization constants to avoid extra kernels.

Reply:
"we could get benefit from larger leaf"
Actually, no... at least for GPU (PVC in particular)...  Even leaf is 4 gives a worse result...
Also for merging stages, 4 is optimal size (for GPU).

Reply:
OK thanks for checking.  

Its a surprise to me that a few extra comparisons without added synchronization in the leaf phase wouldn't be cheaper than an added kernel launch and pass through the data. However, we should of course should follow the data.